locator/topology.cc

```diff
@@ -25,45 +25,375 @@ thread_local const endpoint_dc_rack endpoint_dc_rack::default_location = {
+        ret._nodes_by_host_id = _nodes_by_host_id;
+        ret._nodes_by_endpoint = _nodes_by_endpoint;
+        ret._dc_nodes = _dc_nodes;
+        ret._dc_rack_nodes = _dc_rack_nodes;
+        ret._dc_endpoints = _dc_endpoints;
+        ret._dc_racks = _dc_racks;
+        ret._datacenters = _datacenters;
+    } else {
+        tlogger.debug("topology[{}]: clone_gently to {} from shard {}", fmt::ptr(this), fmt::ptr(&ret), _shard);
+        for (const auto& n : _all_nodes) {
+            ret.add_node(make_lw_shared<node>(*n));
+            co_await coroutine::maybe_yield();
+        }
+        // local node may be detached for _all_nodes
+        // if it was decommissioned.
+        if (!ret._local_node && _local_node) {
+            ret.add_node(make_lw_shared<node>(*_local_node));
         }
     }
     co_await coroutine::maybe_yield();
-    ret._current_locations.reserve(_current_locations.size());
-    for (const auto& p : _current_locations) {
-        ret._current_locations.emplace(p);
-    }
-    co_await coroutine::maybe_yield();
-    ret._datacenters = _datacenters;
     ret._sort_by_proximity = _sort_by_proximity;
     co_return ret;
 }
 
+node_ptr topology::add_node(host_id id, const inet_address& ep, const endpoint_dc_rack& dr) {
+    if (dr.dc.empty() || dr.rack.empty()) {
+        on_internal_error(tlogger, "Node must have valid dc and rack");
+    }
+    auto is_local = node::local(ep == utils::fb_utilities::get_broadcast_address());
+    if (is_local && _local_node) {
+        if (_local_node->host_id() == id) {
+            on_internal_error_noexcept(tlogger, format("Local node already set: host_id={} endpoint={} dc={} rack={}: currently mapped to host_id={} endpoint={}",
+                    id, ep, dr.dc, dr.rack,
+                    _local_node->host_id(), _local_node->endpoint()));
+            return _local_node;
+        }
+        // Replacing node with the same ip address
+        is_local = node::local::no;
+    }
+    return add_node(make_lw_shared<node>(id, ep, dr, is_local));
+}
+
+node_ptr topology::add_node(mutable_node_ptr node) {
+    tlogger.debug("topology[{}]: add_node: node={} host_id={} endpoint={} dc={} rack={} local={}, at {}", fmt::ptr(this), fmt::ptr(node.get()),
+            node->host_id(), node->endpoint(), node->dc_rack().dc, node->dc_rack().rack, node->is_local(), current_backtrace());
+    if (node->is_local() && _local_node && _local_node != node) {
+        on_internal_error(tlogger, format("Local node already set: host_id={} endpoint={} dc={} rack={}: currently mapped to host_id={} endpoint={}",
+                node->host_id(), node->endpoint(), node->dc_rack().dc, node->dc_rack().rack,
+                _local_node->host_id(), _local_node->endpoint()));
+    }
+    // Note that _all_nodes contains also the _lock_node
+    if (_all_nodes.contains(node)) {
+        return node;
+    }
+    try {
+        // FIXME: for now we allow adding nodes with null host_id
+        if (node->host_id()) {
```

Comment:
Let's not allow it. Or, if we do allow it, let's not allow changing it; instead create a new topology.

Reply:
We must allow it for _this_node as part of bootstrapping the topology since when topology is constructed we don't know yet the local host id or it may not even be created for a pristine node, changing that will require significant restructuring that I'd like to avoid in this branch.
If it's indeed the only case, I think that restricting insertion of nodes with null host_id to this node is reasonable.

Reply:
I see. We really should hack a way to read the local id ahead of time. Since it's immutable, we can just construct a reader from the LOCAL table's sstables.

Reply:
or keep it in scylla.conf